src/pipeline.ts

```
@@ -0,0 +1,166 @@
+}
+
+/** Create a new Snowpack Pipeline from a config. */
+export default class Pipeline {
+  config: SnowpackConfig;
+  isDev: boolean;
+
+  constructor({config, isDev}: {config: SnowpackConfig; isDev?: boolean}) {
+    this.config = config;
+    this.isDev = isDev || false;
+  }
+
+  /** Load a Snowpack plugin from a string or tuple */
+  loadPlugin(name: string, options: any = {}): SnowpackPlugin | undefined {
+    const config = this.config;
+
+    // internal proxy plugin (remove this when published to npm)
+    if (name === '@snowpack/plugin-proxy') {
+      return require('./plugin-proxy');
+    }
+
+    try {
+      const pluginLoc = require.resolve(name, {paths: [process.cwd()]});
+      return require(pluginLoc)(config, options);
+    } catch (err) {
+      return undefined;
+    }
+  }
+
+  /** Match a file with an existing pipeline */
+  public match(
+    filename: string,
+  ): {matched: string; replace?: [string, string]; pipeline: SnowpackPipelineStep[]} {
+    const pipeline = this.config.pipeline as SnowpackPipeline;
+
+    // note: don’t use path.extname() because that will convert things like *.proxy.js to .js
+    const allMatchers = Object.keys(pipeline);
+
+    for (let i = 0; i < allMatchers.length; i++) {
+      const matchedFiles = parseExtString(allMatchers[i]);
+      const firstMatch = matchedFiles.find(({input}) => filename.endsWith(input));
+
+      // short-circuit loop as soon as a match is found and don’t continue
+      if (firstMatch) {
+        const {input, output} = firstMatch;
+        return {
+          matched: input,
+          pipeline: pipeline[allMatchers[i]],
+          replace: output ? [input, output] : undefined,
+        };
+      }
+    }
+
+    throw new Error(
+      `Couldn’t find a match in config.pipeline for ${filename}. Add a Snowpack plugin that will handle this for you, or your own filetype matcher for "${path.extname(
+        filename,
+      )}" (see snowpack.dev for more)`,
+    );
+  }
```

🤔 Now that I think about it, should we throw an error if there’s no matcher? This would mean that Snowpack must have pipelines defined for every single type of file, which may be too much.